geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/geotools/GeoMesaFeatureSource.scala

```diff
@@ -59,7 +60,12 @@ class GeoMesaFeatureSource(val ds: DataStore with HasGeoMesaStats,
     import org.locationtech.geomesa.index.conf.QueryProperties.QueryExactCount
 
     val useExactCount = query.getHints.isExactCount.getOrElse(QueryExactCount.get.toBoolean)
-    val count = ds.stats.getCount(getSchema, query.getFilter, useExactCount).getOrElse(-1L)
+
+    val count = if (query.getMaxFeatures < 1000) {
+      SelfClosingIterator(getFeatures(query)).size
+    } else {
+      ds.stats.getCount(getSchema, query.getFilter, useExactCount).getOrElse(-1L)
```

Comment:
we should have a math.min here right?

Reply:
Technically yes.  That'd avoid a situation where GM estimates a count larger than the maxFeatures.